File: talleres/taller8/main.py
import sys

def dfs(v, p, aristas, visitados, backedge, valor_paloma):
    visitados[v] = visitados[p] + 1
    cubren = 0
    cubren_hijo = 0

    for u in aristas[v]:
        if(not visitados[u]):
            cubren_hijo = dfs(u , v, aristas, visitados, backedge, valor_paloma)

            if(cubren_hijo - backedge[v] == 0):
                valor_paloma[v] += 1
            
            cubren += cubren_hijo - backedge[v]
            backedge[v] = 0
        elif(u != p and visitados[u] < visitados[v]):
            cubren += 1
            backedge[u] += 1
    
    return cubren
    

def mergeSort(lista):
    if(len(lista) > 1):
        mitad = len(lista) // 2

        mitad_izq = lista[:mitad]
        mitad_der = lista[mitad:]
        mergeSort(mitad_izq)
        mergeSort(mitad_der)

        i = j = k = 0

        while(i < len(mitad_izq) and j < len(mitad_der)):
            if(mitad_izq[i][1] > mitad_der[j][1]):
                lista[k] = mitad_izq[i]
                i+= 1
            elif(mitad_izq[i][1] == mitad_der[j][1]):
                if(mitad_izq[i][0] < mitad_der[j][0]):
                    lista[k] = mitad_izq[i]
                    i+= 1
                else:
                    lista[k] = mitad_der[j]
                    j += 1
            else:
                lista[k] = mitad_der[j]
                j += 1
            k += 1
        
        while(i < len(mitad_izq)):
            lista[k] = mitad_izq[i]
            i += 1
            k += 1
        
        while(j < len(mitad_der)):
            lista[k] = mitad_der[j]
            j += 1
            k += 1


datos = input().split()
N = int(datos[0])
M = int(datos[1])
while(N != 0 and M != 0):
    cant_estaciones = N
    estaciones_candidatas = M
    aristas = [[] for _ in range(int(cant_estaciones))]
    visitados = [0 for _ in range(cant_estaciones)]
    backedge = [0 for _ in range(cant_estaciones)]
    valor_paloma = [1 for _ in range(cant_estaciones)]
    valor_paloma[0] = 0

    datos = input().split()
    v = int(datos[0])
    u = int(datos[1])
    while(v != -1 and u != -1):
        aristas[v].append(u)
        aristas[u].append(v)
        datos = input().split()
        v = int(datos[0])
        u = int(datos[1])
    
    dfs(0, -1, aristas, visitados, backedge, valor_paloma)

    for i in range(cant_estaciones):
        valor = valor_paloma[i]
        valor_paloma[i] = (i, valor)

    mergeSort(valor_paloma)

    for i in range(estaciones_candidatas):
        print(str(valor_paloma[i][0]) + " " + str(valor_paloma[i][1]))
    print(" ")
    datos = input().split()
    N = int(datos[0])
    M = int(datos[1])









    

File: talleres/taller8/Main.java

import java.util.ArrayList;
import java.util.Scanner;
public class Main {
    public static int dfs(int v, int p, ArrayList<ArrayList<Integer>> grafo, int[] visitados, int[] backedge, int[] valor_paloma){
        if(p == -1){
            visitados[v] = 1;
        } else{
            visitados[v] = visitados[p] + 1;
        }
        int cubren = 0;
        int cubren_hijo = 0;

        for(int i = 0; i < grafo.get(v).size(); i++){
            int hijo = grafo.get(v).get(i);
            if(visitados[hijo] == 0){
                cubren_hijo = dfs(hijo, v, grafo, visitados, backedge, valor_paloma);

                if(cubren_hijo - backedge[v] == 0){
                    valor_paloma[v]++;
                }
                cubren += cubren_hijo - backedge[v];
                backedge[v] = 0;
            }
            else if(hijo != p && visitados[hijo] < visitados[v]){
                cubren++;
                backedge[hijo]++;
            }
        }
        return cubren;
    }

    public static void sort(int[][] lista, int izquierda, int derecha){
        if(izquierda < derecha){
            int mitad = (izquierda + derecha) / 2;

            sort(lista, izquierda, mitad);
            sort(lista, mitad+1, derecha);

            merge(lista, izquierda, mitad, derecha);
        }
    }

    public static void merge(int[][] lista, int izquierda, int mitad, int derecha){
        int n1 = mitad - izquierda + 1;
        int n2 = derecha - mitad;

        int [][] mitad_izquierda = new int[n1][];
        int [][] mitad_derecha = new int[n2][];

        for(int i = 0; i < n1; i++){
            mitad_izquierda[i] = lista[izquierda+i];
        }
        for(int j = 0; j < n2; j++){
            mitad_derecha[j] = lista[mitad + j + 1];
        }

        int i = 0; int j = 0; int k = izquierda;

        while(i < n1 && j < n2){
            if(mitad_izquierda[i][1] > mitad_derecha[j][1]){
                lista[k] = mitad_izquierda[i];
                i++;
            } else if(mitad_izquierda[i][1] == mitad_derecha[j][1]){
                if(mitad_izquierda[i][0] < mitad_derecha[j][0]){
                    lista[k] = mitad_izquierda[i];
                    i++;
                } else{
                    lista[k] = mitad_derecha[j];
                    j++;
                }
            } else{
                lista[k] = mitad_derecha[j];
                j++;
            } 
            k++;
        }

        while (i < n1) {
            lista[k] = mitad_izquierda[i];
            i++;
            k++;
          }

          while (j < n2) {
            lista[k] = mitad_derecha[j];
            j++;
            k++;
          }
    }


    public static void main(String[] args){
        Scanner scanner = new Scanner(System.in);
        int N = scanner.nextInt();
        int M = scanner.nextInt();
        while(N != 0 && M != 0){
            int cant_estaciones = N;
            int estaciones_candidatas = M;
            ArrayList<ArrayList<Integer>> grafo = new ArrayList<>();
            for(int k = 0; k < cant_estaciones; k++){
                grafo.add(new ArrayList<>());
            }
            int[] visitados = new int[cant_estaciones];
            int[] backedge = new int[cant_estaciones];
            int[] valor_paloma = new int[cant_estaciones];
            for(int i = 1; i < cant_estaciones; i++){
                valor_paloma[i] = 1;
            }

            int v = scanner.nextInt();
            int u = scanner.nextInt();
            while(v != -1 && u != -1){
                grafo.get(v).add(u);
                grafo.get(u).add(v);
                v = scanner.nextInt();
                u = scanner.nextInt();
            }
            dfs(0, -1, grafo, visitados, backedge, valor_paloma);
            int[][] valor_paloma_index = new int[cant_estaciones][2];
            for(int j = 0; j < cant_estaciones; j++){
                valor_paloma_index[j][0] = j;
                valor_paloma_index[j][1] = valor_paloma[j];
            }
            sort(valor_paloma_index, 0, valor_paloma_index.length-1);
            for(int j = 0; j < estaciones_candidatas; j++){
                System.out.println(valor_paloma_index[j][0] + " " + valor_paloma_index[j][1]);
            }
            N = scanner.nextInt();
            M = scanner.nextInt();
            System.out.println();
        }
        scanner.close();
    }
}

